File: src/main/scala/com/imcode/imcms/api/doc_getter_callback.scala
package com.imcode
package imcms.api

import com.imcode.imcms.mapping.DocumentMapper
import imcode.server.document.DocumentDomainObject
import imcode.server.user.UserDomainObject
import javax.servlet.http.HttpServletRequest
import imcode.server.{ImcmsServices, ImcmsConstants}


object DocGetterCallbacks {

  private sealed trait DocVersionType
  private case object WorkingVersion extends DocVersionType
  private case object DefaultVersion extends DocVersionType
  private case class CustomVersion(no: Int) extends DocVersionType

  private object DocVersionType {
    def unapply(string: String): Option[DocVersionType] =
      PartialFunction.condOpt(string.trimToEmpty.toLowerCase) {
        case ImcmsConstants.REQUEST_PARAM_VALUE__DOC_VERSION__ALIAS_DEFAULT => DefaultVersion
        case ImcmsConstants.REQUEST_PARAM_VALUE__DOC_VERSION__ALIAS_WORKING => WorkingVersion
        case NonNegInt(no) if no == DocumentVersion.WORKING_VERSION_NO => WorkingVersion
        case NonNegInt(no) => CustomVersion(no)
      }
  }

  /** Creates a callback and sets it to the user. */
  def updateUserDocGetterCallback(request: HttpServletRequest, services: ImcmsServices, user: UserDomainObject) {
    val currentDocGetterCallback = user.getDocGetterCallback
    val docI18nSupport = services.getDocumentLanguageSupport
    val defaultLanguage = docI18nSupport.getDefaultLanguage
    val preferredLanguage = request.getParameter(ImcmsConstants.REQUEST_PARAM__DOC_LANGUAGE).trimToOption
                   .flatMap(code => docI18nSupport.getByCode(code).asOption)
                   .orElse(currentDocGetterCallback.asOption.map(_.documentLanguages.preferred))
                   .orElse(docI18nSupport.getForHost(request.getServerName).asOption)
                   .getOrElse(defaultLanguage)

    val documentLanguages = DocumentLanguages(preferredLanguage, defaultLanguage)
    val docGetterCallback =
      (for {
        NonNegInt(docId) <- request.getParameter(ImcmsConstants.REQUEST_PARAM__DOC_ID).asOption
        DocVersionType(docVersionType) <- request.getParameter(ImcmsConstants.REQUEST_PARAM__DOC_VERSION).asOption
        if !user.isDefaultUser
      } yield {
        docVersionType match {
          case WorkingVersion => WorkingDocGetterCallback(documentLanguages, docId)
          case DefaultVersion => DefaultDocGetterCallback(documentLanguages)
          case CustomVersion(docVersionNo) => CustomDocGetterCallback(documentLanguages, docId, docVersionNo)
        }
      }) getOrElse {
        currentDocGetterCallback match {
          case customDocGetterCallback: CustomDocGetterCallback => customDocGetterCallback.copy(documentLanguages)
          case workingDocGetterCallback: WorkingDocGetterCallback => workingDocGetterCallback.copy(documentLanguages)
          case _ => DefaultDocGetterCallback(documentLanguages)
        }
      }

    user.setDocGetterCallback(docGetterCallback)
  }
}


/**
 * Parametrized callback for DocumentMapper#getDocument method.
 * A callback is (re)created on each request and (re)assigned to a user.
 *
 * Default doc callback always returns default version of any doc if it is present and the user has at least 'view' rights on it.
 *
 * Working and Custom doc callback return working and custom version of a document with particular id;
 * for other doc ids they behave exactly as default doc callback.
 *
 * @see imcode.server.Imcms
 * @see com.imcode.imcms.servlet.ImcmsFilter
 * @see com.imcode.imcms.mapping.DocumentMapper#getDocument(Integer)
 */
trait DocGetterCallback {
  def documentLanguages: DocumentLanguages
  def getDoc[A <: DocumentDomainObject](docId: Int, user: UserDomainObject, docMapper: DocumentMapper): A
}

case class DefaultDocGetterCallback(documentLanguages: DocumentLanguages) extends DocGetterCallback {
  def getDoc[A <: DocumentDomainObject](docId: Int, user: UserDomainObject, docMapper: DocumentMapper): A =
    docMapper.getDefaultDocument[A](docId, documentLanguages.preferred) match {
      case doc if doc != null && !documentLanguages.preferredIsDefault && user.isSuperAdmin =>
        val meta = doc.getMeta

        if (!meta.getEnabledLanguages.contains(documentLanguages.preferred)) {
          if (meta.getDisabledLanguageShowSetting == Meta.DisabledLanguageShowSetting.SHOW_IN_DEFAULT_LANGUAGE)
            docMapper.getDefaultDocument(docId, documentLanguages.default)
          else
            null.asInstanceOf[A]
        } else doc

      case doc => doc
    }
}

case class WorkingDocGetterCallback(documentLanguages: DocumentLanguages, selectedDocId: Int) extends DocGetterCallback {
  def getDoc[A <: DocumentDomainObject](docId: Int, user: UserDomainObject, docMapper: DocumentMapper): A =
    if (selectedDocId == docId) docMapper.getWorkingDocument(docId, documentLanguages.preferred)
    else DefaultDocGetterCallback(documentLanguages).getDoc(docId, user, docMapper)
}

case class CustomDocGetterCallback(documentLanguages: DocumentLanguages, selectedDocId: Int, selectedDocVersionNo: Int) extends DocGetterCallback {
  def getDoc[A <: DocumentDomainObject](docId: Int, user: UserDomainObject, docMapper: DocumentMapper): A =
    if (selectedDocId == docId) docMapper.getCustomDocument(DocRef.of(selectedDocId, selectedDocVersionNo, documentLanguages.preferred.getCode))
    else DefaultDocGetterCallback(documentLanguages).getDoc(docId, user, docMapper)
}

case class DocumentLanguages(preferred: DocumentLanguage, default: DocumentLanguage) {
  val preferredIsDefault = preferred == default
}

File: src/main/java/com/imcode/imcms/mapping/DocGetterCallback.java
package com.imcode.imcms.mapping;

import com.imcode.imcms.api.DocumentVersion;
import com.imcode.imcms.mapping.container.DocRef;
import com.imcode.imcms.model.Language;
import imcode.server.Imcms;
import imcode.server.document.DocumentDomainObject;
import imcode.server.user.UserDomainObject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

/**
 * Parametrized callback for DocumentMapper#getDocument method.
 * A callback is updated on each request and (re)assigned to a user.
 * <p>
 * Default doc callback always returns default version of any doc if it is present and the user has at least 'view' rights on it.
 * <p>
 * Working and Custom doc callback return working and custom version of a document with particular id;
 * for other doc ids they behave exactly as default doc callback.
 *
 * @see imcode.server.Imcms
 * @see com.imcode.imcms.servlet.ImcmsSetupFilter
 * @see com.imcode.imcms.mapping.DocumentGetter#getDocument(int)
 */
public class DocGetterCallback implements Serializable {

    private static final Logger logger = LoggerFactory.getLogger(DocGetterCallback.class);
    private static final long serialVersionUID = 2496394918087427549L;
    private volatile Language language;
    private UserDomainObject user;
    private Map<Integer, Callback> callbacks = new ConcurrentHashMap<>();
    private final Callback workingDocCallback = (docId, docMapper) -> {
        logger.trace("Working doc requested - user: {}, docId: {}, language: {}.", user, docId, language);
        return docMapper.getWorkingDocument(docId, language);
    };
    private final Callback uncheckedDefaultDocCallback = (docId, docMapper) -> {
        logger.trace("Default doc (unchecked) requested - user: {}, docId: {}, language: {}.", user, docId, language);
        return docMapper.getDefaultDocument(docId, language);
    };
    private final Callback defaultDocCallback = (docId, docMapper) -> {
        logger.trace("Default doc requested - user: {}, docId: {}, language: {}.", user, docId, language);

        DocumentDomainObject doc = docMapper.getDefaultDocument(docId, language);

        if (doc != null) {
            List<Language> docLanguages = Imcms.getServices()
                    .getDocumentMapper()
                    .getCommonContents(doc.getId(), doc.getVersionNo())
                    .entrySet()
                    .stream()
                    .filter(langToContent -> langToContent.getValue().getEnabled())
                    .map(Map.Entry::getKey)
                    .collect(Collectors.toCollection(ArrayList::new));

            if (docLanguages.isEmpty()) {
                docLanguages.add(doc.getLanguage());
            }

            if (!docLanguages.contains(language)) { // current language is disabled for current document
                doc = shouldDocBeShownWithDefaultLang(doc, docLanguages)
                        ? docMapper.getDefaultDocument(docId)
                        : null;
            }
        }

        return doc;
    };

    public DocGetterCallback(UserDomainObject user) {
        this.user = user;
    }

    private boolean shouldDocBeShownWithDefaultLang(DocumentDomainObject doc, List<Language> docLanguages) {
        return (doc.getDisabledLanguageShowMode() == DocumentMeta.DisabledLanguageShowMode.SHOW_IN_DEFAULT_LANGUAGE
                && docLanguages.contains(Imcms.getServices().getLanguageService().getDefaultLanguage()));
    }

    @SuppressWarnings("unchecked")
    public <T extends DocumentDomainObject> T getDoc(int docId, DocumentMapper docMapper) {
        Callback callback = user.isDefaultUser()
                ? defaultDocCallback
                : callbacks.getOrDefault(docId, uncheckedDefaultDocCallback);

        return (T) callback.getDoc(docId, docMapper);
    }

    public void setDefault(int docId) {
        callbacks.remove(docId);
    }

    public void setWorking(int docId) {
        if (!user.isDefaultUser()) {
            callbacks.put(docId, workingDocCallback);
        }
    }

    public void setCustom(int docId, int versionNo) {
        if (!user.isDefaultUser()) {
            callbacks.put(docId, versionNo == DocumentVersion.WORKING_VERSION_NO ? workingDocCallback : createCustomDocCallback(versionNo));
        }
    }

    private Callback createCustomDocCallback(int versionNo) {
        return (docId, docMapper) -> {
            logger.trace("Custom doc requested - user: {},  docId: {}, versionNo: {}, language: {}.", user, docId, versionNo, language);
            return docMapper.getCustomDocument(DocRef.of(docId, versionNo, language.getCode()));
        };
    }

    public Language getLanguage() {
        return language;
    }

    public void setLanguage(Language language) {
        this.language = language;
    }

    private interface Callback extends Serializable {
        DocumentDomainObject getDoc(int docId, DocumentMapper docMapper);
    }
}
